### PR TITLE
Cast from integer/timestamp to timestamp/integer

### LINKED ISSUE #5039: Cast from integer/timestamp to timestamp/integer
**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**

Only Int64 can be casted to/from Timestamp for now. In Spark, other integer types can be casted to/from Timestamp type as like Int64 (i.e., LongType). We'd like to fill this gap in cast kernel too.

**Describe the solution you'd like**

Supporting casting from integer/timestamp to timestamp/integer by cast kernel.

**Describe alternatives you've considered**

**Additional context**

### PR DESCRIPTION
**Pull request by viirya**
# Which issue does this PR close?

Closes #5039.

# Rationale for this change
 

# What changes are included in this PR?

# Are there any user-facing changes?

### CODE REVIEW

File: arrow-cast/src/cast.rs [hunk 1]
```diff
@@ -1621,24 +1621,104 @@ pub fn cast_with_options(
...
+                    from_type, to_type
+                ))),
+            }
         }
-        (Timestamp(TimeUnit::Microsecond, _), Int64) => {
-            cast_reinterpret_arrays::<TimestampMicrosecondType, Int64Type>(array)
+        (Timestamp(TimeUnit::Microsecond, _), _) if to_type.is_integer() => {
+            let array = cast_reinterpret_arrays::<TimestampMicrosecondType, Int64Type>(array)?;
+
+            match to_type {
```
Comment (tustvold) [inline arrow-cast/src/cast.rs:1663]: I wonder if instead of this additional match block we could call the kernels recursively?

**Review by tustvold — APPROVED** [review]
This looks good, although I wonder if calling the kernel recursively might be simpler (and possibly less codegen)

File: arrow-cast/src/cast.rs (hunk 1 above)
Comment (viirya) [reply arrow-cast/src/cast.rs]: Yea, it's simpler. 👍

**Comment by viirya** [comment]
Thank you @tustvold

**Pull request merged**